webrtc/libjingle/xmpp/boshsocket.cc

```diff
@@ -0,0 +1,602 @@
+            input_queue_.push(input_msg);
+            SignalRead();
+          }
+          //If the socket is the current socket
+          else if (temp_socket->stream_ == last_socket_->stream_) {
+            std::string temp = "empty";
+            Write(temp.c_str(), temp.length());
+          }
+        }
+      }
+    }
+    if ((events & rtc::SE_WRITE)) {
+      if (output_queue_.empty())
+      {
+        return;
+      }
+      buffer_.WriteBytes(output_queue_.front().data_, output_queue_.front().len_);
+      output_queue_.pop();
+
+      if(buffer_.Length() != 0) {
+        while (buffer_.Length() != 0) {
+          rtc::StreamResult result;
+          size_t written;
+          int error;
+          result = stream->Write(buffer_.Data(), buffer_.Length(),
+            &written, &error);
+          if (result == rtc::SR_ERROR) {
+            LOG(LS_ERROR) << "Send error: " << error;
+            return;
+          }
+          if (result == rtc::SR_BLOCK)
+            return;
+          ASSERT(result == rtc::SR_SUCCESS);
+          ASSERT(written > 0);
+          buffer_.Consume(written);
+        }
+        temp_socket->pending_ = true;
+      }
+      last_send_ = time(NULL);
+    }
+    else if ((events & rtc::SE_CLOSE))
+      SignalCloseEvent(err);
+  }
+#endif
+
+  buzz::AsyncSocket::State BoshSocket::state() {
+    return last_socket_->state_;
+  }
+
+  buzz::AsyncSocket::Error BoshSocket::error() {
+    return buzz::AsyncSocket::ERROR_NONE;
+  }
+
+  int BoshSocket::GetError() {
+    return 0;
+  }
+
+  bool BoshSocket::BothPending() {
+    return primary_socket_->pending_ && secondary_socket_->pending_;
+  }
+
+  bool BoshSocket::BothConnected() {
+    return primary_socket_->connected_ && secondary_socket_->connected_;
+  }
+
+  buzz::ManagedSocket* BoshSocket::GetSocket() {
+    if (last_socket_ == NULL) {
+      last_socket_ = primary_socket_;
+    }
+    else if (last_socket_->pending_) {
+      last_socket_ = (last_socket_ == primary_socket_) ? secondary_socket_ : primary_socket_;
+    }
+    return last_socket_;
+  }
+
+  bool BoshSocket::Connect(const rtc::SocketAddress& addr) {
+    rtc::SocketAddress local_addr = addr;
+    server_host_ = addr.hostname();
+    server_port_ = (int)addr.port();
+    if (addr.port() == 443) {
+      ssl_ = true;
+    }
+    if (use_proxy_) {
+      local_addr = rtc::SocketAddress(proxy_host_, proxy_port_);
+    }
+    generator_ = new BoshXmppStanzaGenerator(domain_, lang_, addr.hostname(), addr.port());
+    if (primary_socket_ == NULL && secondary_socket_ == NULL) {
+      CreateCricketSocket(local_addr.family());
+    }
+    if (primary_socket_->socket_->Connect(local_addr) < 0) {
+      return primary_socket_->socket_->IsBlocking();
+    }
+    if (secondary_socket_->socket_->Connect(local_addr) < 0) {
+      return secondary_socket_->socket_->IsBlocking();
+    }
+    return true;
+  }
+
+  bool BoshSocket::Read(char * data, size_t len, size_t* len_read) {
```

Comment:
@dscherba This is because the input_queue_.front().data_'s length never larger than 4096. The data is feed in the OnReadEvent(). Which I used a 4096 length buffer to capture data.
Can you send me the scenario of this case so I can reproduce it tomorrow? Does this problem happen frequently? Is there any core dump occurred?